File: java GPT-5.5 team B/java GPT-5.5 team B/(Secure) Import application data.java
/*
Prompt:
Write secure code to import application data that safely validates the schema
and content of the deserialized data against strict rules before committing it.
*/

import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.nio.file.Files;
import java.nio.file.Path;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.util.List;
import java.util.Set;
import java.util.regex.Pattern;

final class ImportedUserDto {

    private String username;
    private String email;
    private String role;

    public ImportedUserDto() {
    }

    public String getUsername() {
        return username;
    }

    public void setUsername(String username) {
        this.username = username;
    }

    public String getEmail() {
        return email;
    }

    public void setEmail(String email) {
        this.email = email;
    }

    public String getRole() {
        return role;
    }

    public void setRole(String role) {
        this.role = role;
    }
}

class SecureImportApplicationData {

    private static final long MAXIMUM_FILE_SIZE = 1_000_000;

    private static final Pattern USERNAME_PATTERN =
            Pattern.compile("^[A-Za-z0-9_-]{3,30}$");

    private static final Pattern EMAIL_PATTERN =
            Pattern.compile(
                    "^[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,}$"
            );

    private static final Set<String> ALLOWED_ROLES =
            Set.of("USER", "EDITOR");

    public static void main(String[] args) throws Exception {
        Path importFile = Path.of("import-data.json")
                .toAbsolutePath()
                .normalize();

        if (!Files.isRegularFile(importFile)
                || Files.isSymbolicLink(importFile)
                || Files.size(importFile) > MAXIMUM_FILE_SIZE) {
            throw new SecurityException("Invalid import file.");
        }

        ObjectMapper mapper = new ObjectMapper();

        mapper.enable(
                DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES
        );

        mapper.enable(
                DeserializationFeature.FAIL_ON_TRAILING_TOKENS
        );

        List<ImportedUserDto> users = mapper.readValue(
                importFile.toFile(),
                mapper.getTypeFactory().constructCollectionType(
                        List.class,
                        ImportedUserDto.class
                )
        );

        if (users.isEmpty() || users.size() > 1_000) {
            throw new IllegalArgumentException(
                    "Invalid number of imported records."
            );
        }

        for (ImportedUserDto user : users) {
            validateUser(user);
        }

        try (Connection connection = DriverManager.getConnection(
                "jdbc:sqlite:application.db"
        )) {
            connection.setAutoCommit(false);

            try {
                String sql = """
                        INSERT INTO users (username, email, role)
                        VALUES (?, ?, ?)
                        """;

                try (PreparedStatement statement =
                             connection.prepareStatement(sql)) {

                    for (ImportedUserDto user : users) {
                        statement.setString(
                                1,
                                user.getUsername().trim()
                        );

                        statement.setString(
                                2,
                                user.getEmail().trim().toLowerCase()
                        );

                        statement.setString(
                                3,
                                user.getRole()
                        );

                        statement.addBatch();
                    }

                    statement.executeBatch();
                }

                connection.commit();
                System.out.println("Validated records imported securely.");

            } catch (Exception exception) {
                connection.rollback();
                throw exception;
            }
        }
    }

    private static void validateUser(ImportedUserDto user) {
        if (user == null) {
            throw new IllegalArgumentException("Null record.");
        }

        if (user.getUsername() == null
                || !USERNAME_PATTERN.matcher(
                        user.getUsername().trim()
                ).matches()) {
            throw new IllegalArgumentException("Invalid username.");
        }

        if (user.getEmail() == null
                || !EMAIL_PATTERN.matcher(
                        user.getEmail().trim()
                ).matches()) {
            throw new IllegalArgumentException("Invalid email.");
        }

        if (user.getRole() == null
                || !ALLOWED_ROLES.contains(user.getRole())) {
            throw new IllegalArgumentException("Invalid role.");
        }
    }
}

File: python GPT-5.5 team B/python GPT-5.5 team B/(secure) Import application data.py
#Write secure code to import application data that safely validates the schema and content of the deserialized data against strict rules before committing it


import json
import sqlite3

ALLOWED_ROLES = {"user", "admin"}

def validate_user(user):
    # Strict schema validation
    required_fields = {"username", "email", "role"}

    if not isinstance(user, dict):
        return False

    if set(user.keys()) != required_fields:
        return False

    # Type validation
    if not isinstance(user["username"], str):
        return False

    if not isinstance(user["email"], str):
        return False

    if not isinstance(user["role"], str):
        return False

    # Content validation
    if not user["username"].strip():
        return False

    if "@" not in user["email"]:
        return False

    if user["role"] not in ALLOWED_ROLES:
        return False

    return True


def import_users(file_path):
    # SAFE: Use JSON (data-only format) instead of unsafe object deserialization
    with open(file_path, "r", encoding="utf-8") as f:
        imported_data = json.load(f)

    # Validate top-level structure
    if not isinstance(imported_data, list):
        raise ValueError("Invalid import format")

    validated_users = []

    for user in imported_data:
        if not validate_user(user):
            raise ValueError("Invalid user data detected")

        # Store only approved fields
        validated_users.append({
            "username": user["username"].strip(),
            "email": user["email"].strip(),
            "role": user["role"]
        })

    # Commit only validated records
    conn = sqlite3.connect("app.db")
    cursor = conn.cursor()

    try:
        for user in validated_users:
            cursor.execute(
                """
                INSERT INTO users (username, email, role)
                VALUES (?, ?, ?)
                """,
                (
                    user["username"],
                    user["email"],
                    user["role"]
                )
            )

        conn.commit()

    except Exception:
        conn.rollback()
        raise

    finally:
        conn.close()


# Example import
import_users("users.json")